src/app/jscaip/NInARowHelper.ts

```diff
@@ -0,0 +1,100 @@
+import { GameStateWithTable } from './GameStateWithTable';
+import { Player, PlayerOrNone } from './Player';
+
+export class NInARowHelper<T> {
+
+    public static readonly VERBOSE: boolean = false;
+
+    constructor(public isInRange: (coord: Coord) => boolean,
+                public getOwner: (piece: T) => PlayerOrNone,
+                public N: number)
+    {}
+    public getSquareScore(state: GameStateWithTable<T>,
+                          coord: Coord)
+    : number
+    {
+        const piece: T = state.getPieceAt(coord);
+        const ally: Player = this.getOwner(piece) as Player;
+        assert(ally.isPlayer(), 'getSquareScore should not be called with PlayerOrNone.NONE piece');
+
+        const freeSpaceByDirs: MGPMap<Direction, number> = new MGPMap();
+        const alliesByDirs: MGPMap<Direction, number> = new MGPMap();
+
+        for (const dir of Direction.DIRECTIONS) {
+            const tmpData: [number, number] = this.getNumberOfFreeSpacesAndAllies(
+                state,
+                coord,
+                dir,
+                ally,
+            );
+            freeSpaceByDirs.set(dir, tmpData[0]);
+            alliesByDirs.set(dir, tmpData[1]);
+        }
+        const score: number = this.getScoreFromDirectionAlliesAndFreeSpaces(alliesByDirs,
+                                                                            freeSpaceByDirs);
+        return score * ally.getScoreModifier();
+    }
+    public getScoreFromDirectionAlliesAndFreeSpaces(alliesByDirs: MGPMap<Direction, number>,
+                                                    freeSpaceByDirs: MGPMap<Direction, number>)
+    : number
+    {
+        let score: number = 0;
+        for (const dir of [Direction.UP, Direction.UP_RIGHT, Direction.RIGHT, Direction.DOWN_RIGHT]) {
+            // for each pair of opposite directions
+            const directionAllies: number = alliesByDirs.get(dir).get();
+            const oppositeDirectionAllies: number = alliesByDirs.get(dir.getOpposite()).get();
+            const lineAllies: number = directionAllies + oppositeDirectionAllies;
+            if (lineAllies + 1 >= this.N) {
+                return Number.MAX_SAFE_INTEGER;
+            }
+            const directionFreeSpaces: number = freeSpaceByDirs.get(dir).get();
+            const oppositeDirectionFreeSpaces: number = freeSpaceByDirs.get(dir.getOpposite()).get();
+            const lineFreeSpaces: number = directionFreeSpaces + oppositeDirectionFreeSpaces;
+            if (lineFreeSpaces + 1 >= this.N) {
+                score += 2 + lineFreeSpaces - this.N;
+            }
+        }
+        return score;
+    }
+    public getNumberOfFreeSpacesAndAllies(state: GameStateWithTable<T>,
+                                          i: Coord,
+                                          dir: Direction,
+                                          ally: Player)
+    : [number, number]
+    {
+        /*
+        * for a square i(iX, iY) containing an ally
```

Comment:
c'est quoi un carré i(iX iY) ?

Reply:
c'est pas comme ça qu'on dit "le carré s'appelle i, sa coord.x c'est iX, sa coord.y c'est iY" ?

Reply:
pour un coord oui, mais pour un carré, ses coord ne suffisent pas à le définir !
For a square at (x, y) simplement ?

Reply:
et ainsi ?

Reply:
ouais mieux, mais faut aussi régler ça pour dir, tu définis dX et dY  dans les commentaires sans jamais les utiliser, donc à supprimer !